File: 85. Maximal Rectangle.go
package main

import (
	"fmt"
)

func maximalRectangle(matrix [][]byte) int {
	if matrix == nil{
		return 0;
	}
	n := len(matrix[0])
    res := 0;
	hist := make([] int, n);
	for _, row:= range matrix{
		for i := 0; i < n; i++{
			if(row[i] == '0'){
				hist[i] = 0;
			} else{
				hist[i] += 1;
			}
		}
		res = max(res, maxReact(hist));
	}

	return res;
}

func maxReact(heights []int) (res int){
	res = 0;
	stack := [] int {};

	for i := 0; i <= len(heights); i++{

		for len(stack) > 0 && (i == len(heights) || heights[stack[len(stack) - 1]] >= heights[i]){
			h := heights[stack[len(stack) - 1]];
			stack = stack[: len(stack) - 1];
			w := i;
			if(len(stack) > 0){
				w = i - stack[len(stack) - 1] - 1;
			}
			res = max(res, h * w);
		}

		stack = append(stack, i);
	}

	return
}

File: 85. Maximal Rectangle.py
class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if not matrix:
            return 0
        res = 0
        hist = [0]*len(matrix[0])
        for row in matrix:
            for i in range(len(row)):
                hist[i] = 0 if row[i] == 0 else hist[i] + 1
            res = max(res, self.maxReact(hist))
        return res
    
    def maxReact(self, heights: List[int]) -> int:
        res = 0
        stack = []

        for i in range(len(heights) + 1):
            while stack and (i == len(heights) or heights[stack[-1]] >= heights[i]):
                h = heights[stack.pop()]
                w = i if not stack else i - stack[-1] - 1
                res = max(res, h * w)

            stack.append(i)

        return res